authgear/authgear-server
# Implement whatsapp otp

ref #3103 #3102

# How to test

1. In authgear.yaml, add the following section
  ```yaml
  messaging:
    whatsapp:
      api_type: "on-premises"
      templates:
        otp:
          name: authgear_whatsapp_otp
          type: authentication
          namespace: "WATI_NAMESPACE"
          languages:
            - en
            - zh_CN
            - zh_HK
  ```

2. In authgear.secrets.yaml, add the following
  ```yaml
  - data:
      username: "WATI_USERNAME"
      password: "WATI_PASSWORD"
      api_endpoint: "WATI_ENDPOINT"
    key: whatsapp.on-premises
  ```

3. Ensure `authenticator.oob_otp.sms.phone_otp_mode` is `whatsapp_sms` in authgear.yaml.
4. Login / Signup / Add Identity with whatsapp

Find all credentials in 1password.

## Available templates

I've created two templates in our wati whatsapp account

1. authgear_whatsapp_otp
  -  `type` must be `authentication` when using this template.
  - Supported languages are `en` `zh_CN` `zh_HK`
<del>
2. authgear_otp_util
  - `type` must be `util` when using this template.
  - Supported languages is only `en`
  - `components` must be specified to tell authgear server how to use this template. You can use the following config:
    ```
    components:
      header:
        parameters:
          - '{{ template "app.name" }}'
      body:
        parameters:
          - '{{ template "whatsapp.body.otp" }}'
          - "{{ .Code }}"
          - '{{ template "whatsapp.body.action" }}'
    ```
    - You need to add translation strings `whatsapp.body.otp` and `whatsapp.body.action` in `var/templates/<language>/translation.json`. For reference:
      ```json
      {
        "whatsapp.body.otp": "one-time password",
        "whatsapp.body.action": "logging in",
      }
      ```
</del>

TODO:

  - [x] Remove util template support
